File: src/test/java/org/meetkt/catalogue/domain/model/CatalogueUnitTest.java
package org.meetkt.catalogue.domain.model;

import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.meetkt.catalogue.fixture.ProductFixture.aProduct;

class CatalogueUnitTest {

    @Test
    void shouldNotReturnAProductGivenCatalogueIsEmpty() {
        Catalogue catalogue = new Catalogue(Collections.emptyList());

        Optional<Product> product = catalogue.productWith("any-barcode");

        assertThat(product.isEmpty()).isTrue();
    }

    @Test
    void shouldNotReturnAProductGivenProductDoesNotExistForTheBarcode() {
        Catalogue catalogue = new Catalogue(
                List.of(aProduct().withProductId("P-001").withBarcode("barcode-001").build())
        );

        Optional<Product> product = catalogue.productWith("non-existent-product-for-barcode");

        assertThat(product.isEmpty()).isTrue();
    }

    @Test
    void shouldReturnAProductWithProductIdGivenABarcode() {
        Catalogue catalogue = new Catalogue(
                List.of(aProduct().withProductId("P-001").withBarcode("barcode-001").build())
        );

        Optional<Product> product = catalogue.productWith("barcode-001");

        assertThat(product.get().id()).isEqualTo(new ProductId("P-001"));
    }
}

File: src/test/kotlin/in/meetkt/catalogue/domain/model/CatalogueUnitTest.kt
package `in`.meetkt.catalogue.domain.model

import `in`.meetkt.catalogue.fixture.ProductFixture
import arrow.core.getOrElse
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class CatalogueUnitTest {

    @Test
    fun shouldNotReturnAProductGivenCatalogueIsEmpty() {
        val catalogue = Catalogue(emptyList())

        val product = catalogue.productWith("any-barcode")

        assertThat(product.isEmpty()).isTrue
    }

    @Test
    fun shouldNotReturnAProductGivenProductDoesNotExistForTheBarcode() {
        val catalogue =
                Catalogue(listOf(ProductFixture.aProduct().withProductId("001").withBarcode("barcode-001").build()))

        val product = catalogue.productWith("non-existent-product-for-barcode")

        assertThat(product.isEmpty()).isTrue
    }

    @Test
    fun shouldReturnAProductWithProductIdGivenABarcode() {
        val catalogue =
                Catalogue(listOf(ProductFixture.aProduct().withProductId("001").withBarcode("barcode-001").build()))

        val product = catalogue.productWith("barcode-001")

        assertThat(product.getOrElse { ProductFixture.aProduct().build() }.id).isEqualTo(ProductId("001"))
    }
}
